70_FLIPFluids/main.cpp

```diff
@@ -0,0 +1,2240 @@
+		{
+			dispatchUpdateFluidCells(cmdbuf);			// particle to grid
+			dispatchApplyBodyForces(cmdbuf, i == 0);	// external forces, e.g. gravity
+			dispatchApplyDiffusion(cmdbuf);
+			dispatchApplyPressure(cmdbuf);
+			dispatchExtrapolateVelocities(cmdbuf);	// grid -> particle vel
+			dispatchAdvection(cmdbuf);				// update/advect fluid
+		}
+
+		{
+			SMemoryBarrier memBarrier;
+			memBarrier.srcStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+			memBarrier.srcAccessMask = ACCESS_FLAGS::SHADER_WRITE_BITS;
+			memBarrier.dstStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+			memBarrier.dstAccessMask = ACCESS_FLAGS::SHADER_READ_BITS;
+			cmdbuf->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, {.memBarriers = {&memBarrier, 1}});
+        }
+
+		// prepare particle vertices for render
+		cmdbuf->bindComputePipeline(m_genParticleVerticesPipeline.get());
+		cmdbuf->bindDescriptorSets(nbl::asset::EPBP_COMPUTE, m_genParticleVerticesPipeline->getLayout(), 1, 1, &m_genVerticesDs.get());
+		cmdbuf->dispatch(WorkgroupCountParticles, 1, 1);
+
+		{
+			SMemoryBarrier memBarrier;
+			memBarrier.srcStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+			memBarrier.srcAccessMask = ACCESS_FLAGS::SHADER_WRITE_BITS;
+			memBarrier.dstStageMask = PIPELINE_STAGE_FLAGS::VERTEX_INPUT_BITS;
+			memBarrier.dstAccessMask = ACCESS_FLAGS::MEMORY_READ_BITS;
+			cmdbuf->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, {.memBarriers = {&memBarrier, 1}});
+        }
+
+		// draw particles
+		auto* queue = getGraphicsQueue();
+
+		asset::SViewport viewport;
+		{
+			viewport.minDepth = 1.f;
+			viewport.maxDepth = 0.f;
+			viewport.x = 0u;
+			viewport.y = 0u;
+			viewport.width = m_window->getWidth();
+			viewport.height = m_window->getHeight();
+		}
+		cmdbuf->setViewport(0u, 1u, &viewport);
+
+		VkRect2D scissor{
+			.offset = { 0, 0 },
+			.extent = { m_window->getWidth(), m_window->getHeight() }
+		};
+		cmdbuf->setScissor(0u, 1u, &scissor);		
+
+		IGPUCommandBuffer::SRenderpassBeginInfo beginInfo;
+		VkRect2D currentRenderArea;
+		const IGPUCommandBuffer::SClearColorValue clearValue = { .float32 = {0.f,0.f,0.f,1.f} };
+		const IGPUCommandBuffer::SClearDepthStencilValue depthValue = { .depth = 0.f };
+		{
+			currentRenderArea =
+			{
+				.offset = {0,0},
+				.extent = {m_window->getWidth(),m_window->getHeight()}
+			};
+	
+			auto scRes = static_cast<CDefaultSwapchainFramebuffers*>(m_surface->getSwapchainResources());
+			beginInfo =
+			{
+				.framebuffer = scRes->getFramebuffer(m_currentImageAcquire.imageIndex),
+				.colorClearValues = &clearValue,
+				.depthStencilClearValues = &depthValue,
+				.renderArea = currentRenderArea
+			};
+		}
+		cmdbuf->beginRenderPass(beginInfo, IGPUCommandBuffer::SUBPASS_CONTENTS::INLINE);
+
+		cmdbuf->bindGraphicsPipeline(m_graphicsPipeline.get());
+		cmdbuf->bindDescriptorSets(EPBP_GRAPHICS, m_graphicsPipeline->getLayout(), 1, 1, &m_renderDs.get());
+
+		cmdbuf->draw(numParticles * 6, 1, 0, 0);
```

Comment:
should have used `drawIndexed` and an index buffer {0,1,2,2,1,3} + instancing at least, then you'd have 33% less vertex data

But really you should have done all the vertex generation in the vertex shader and not have the compute shader, then you'd have inf% less vertex data

P.S. Non-indexed draws bad, vertex shader runs many more times because no vertex cache.